core/src/impl/Kokkos_Core.cpp

```diff
@@ -995,10 +995,12 @@ void push_finalize_hook(std::function<void()> f) { finalize_hooks.push(f); }
 
 void finalize() { Impl::finalize_internal(); }
 
-void finalize_all() {
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
+KOKKOS_DEPRECATED void finalize_all() {
```

Comment:
If this is to be deprecated, should `Impl::finalize_internal` also be refactored to not have a default argument? Currently, it seems to be only called from `finalize` and `finalize_all`.

Reply:
I pretty much have the same question. What functionality do we need to offer to initialize and finalize individual backends? What are we promising to, say, `Legion`?

Reply:
I dropped the boolean argument from `Implemented::finalize()`.
Cleaning further raises a few questions
https://github.com/kokkos/kokkos/blob/7a413b5640284c0b3489dbd69f790c4951852a35/core/src/impl/Kokkos_Core.cpp#L123-L127
https://github.com/kokkos/kokkos/blob/7a413b5640284c0b3489dbd69f790c4951852a35/core/src/impl/Kokkos_ExecSpaceInitializer.hpp#L56
Presumably they are space we don't know about that inherits from `ExecSpaceInitializerBase` so we need to be careful there.

Reply:
IMHO, it's fair to also drop it from 
```
void ExecSpaceManager::finalize_spaces(const bool all_spaces)
```
since we control that as well but I agree to be careful about going deeper.

Reply:
Done